File: case_studies/tuto/reduction_reinit/reduction_reinit.cpp
#include <optitrust.h>
const int N = 5000;
void reduction_reinit(double* A, double* sum){
 *sum = 0;
 double x[1];
 x[0] = *sum;
 int nb_threads = 10;
  for (int i = 0; i < N; i++) {
    for (int j = 0; j < N; j++) {
      x[0] += A[MINDEX2(N,N,i,j)];
      A[MINDEX2(N,N,i,j)] = 0.;
    }
  }
 *sum = x[0];
}

File: case_studies/tuto/reduction_reinit/reduction_reinit.ml
open Optitrust
open Prelude
let _ = Flags.check_validity := false
let delocalize_sum = Local_arith (Lit_float (typ_f64,0.)  , Binop_add)
let _ = Run.script_cpp (fun _ ->

  !! Loop.fission ~nest_of:2 [cWrite ~lhs:[cVar "A"] (); tBefore];
  !! Omp.parallel [nbMulti; cFor "i"];
  !! Matrix_basic.delocalize ~init_zero:true ~dim:(var_mut "nbThreads") ~index:"j" ~acc_in_place:true ~ops:delocalize_sum  [cFor "j";tAfter] ;
  !! Matrix.delocalize "depositCorners" ~into:"depositThreadCorners" ~indices:["idCell"; "idCorner"]
      ~init_zero:true ~dim:(var_mut "nbThreads") ~index:"idThread" ~acc_in_place:true ~ops:delocalize_sum ~use:(Some (var "idThread"))
      ~labels:["alloc"; ""; "dealloc"] ~alloc_instr ~dealloc_tg:(Some [cTopFunDef ~regexp:true "dealloc.*"; cFor ""])
      [cLabel "core"]
  ;)


